src/components/triggersAndActions.ts

```diff
@@ -0,0 +1,176 @@
+import path from 'path';
+import getAppRootDir from './getAppRoot';
+
+const schedule = require('node-schedule');
+
+export interface DatabaseRecord {
+  [key: string]: any;
+  code: string;
+}
+
+export interface DatabaseResult {
+  rows: DatabaseRecord[];
+}
+
+export interface TriggerPayload {
+  id: number;
+  trigger: string;
+  table: string;
+  record_id: number;
+}
+
+export interface ActionPayload {
+  id: number;
+  code: string;
+  parameters: { [key: string]: any };
+}
+
+export interface Action {
+  code: string;
+  path: string;
+  name: string;
+  trigger: string;
+  condition: { [key: string]: any };
+  parameter_queries: { [key: string]: any };
+}
+
+const pluginFolder = path.join(getAppRootDir(), 'plugins'); //Change this to config file
+
+// Load actions from Database at server startup
+export const loadActions = async function (
+  client: { [key: string]: Function },
+  actionLibrary: { [key: string]: Function }
+) {
+  console.log('Loading Actions from Database...');
+
+  // const actionLibrary: { [key: string]: Function } = {};
+
+  client
+    .query('SELECT code, path, function_name FROM action_plugin')
+    .then((res: DatabaseResult) => {
+      res.rows.forEach((row) => {
+        const action = require(path.join(pluginFolder, row.path));
+        actionLibrary[row.code] = action[row.function_name];
+      });
+    });
+
+  console.log('Actions loaded');
+
+  // return actionLibrary;
+};
+
+// Load scheduled jobs from Database at server startup
+export const loadScheduledActions = async function (
+  client: { [key: string]: Function },
+  actionLibrary: { [key: string]: Function },
+  actionSchedule: any[]
+) {
+  // const actionSchedule: any[] = [];
+
+  console.log('Loading Scheduled jobs');
+  // Load from Database
+  client
+    .query(
+      `SELECT id, action_code, parameters, time_executed FROM action_queue WHERE status = 'Scheduled' ORDER BY time_executed`
+    )
+    .then((res: DatabaseResult) => {
+      res.rows.forEach((action) => {
+        // console.log(action);
+        const date = new Date(action.time_executed);
+        if (date > new Date(Date.now())) {
+          const job = schedule.scheduleJob(date, function () {
+            executeAction(
+              client,
+              {
+                id: action.id,
+                code: action.action_code,
+                parameters: action.parameters,
+              },
+              actionLibrary
+            );
+          });
+          actionSchedule.push(job);
+        } else {
+          // Overdue jobs to be executed immediately
+          console.log('Executing overdue action:', action.action_code, action.parameters);
+          executeAction(
+            client,
+            {
+              id: action.id,
+              code: action.action_code,
+              parameters: action.parameters,
+            },
+            actionLibrary
+          );
+        }
+      });
+    });
+  // return actionSchedule;
+};
```

Comment:
I think I'll continue on Monday from here 😆

Reply:
Okay, sure thing.

Maybe we merge this PR so we've got something start with, and you can do a more thorough review of Triggers and Actions later?

Reply:
Ah, I think it might be good to have a run on the way it is to change to use the Pg util functions in a separated place. Also some missing interface/types make it more difficult to understand. I think it's better to address points of improvement first, otherwise they can be left behind while we rush other things in.